notebooks/linear_regression.livemd

```diff
@@ -0,0 +1,610 @@
+
+<ins>
+<i>
+Hint:
+</i>
+</ins>
+
+<br />
+
+You can try to add another feature by computing clustering on this data set. It might be a sum or power mean of distances to the clusters. We may predict that centroids will be located in San Francisco Bay and Los Angeles Bay. You can also pass population as weights to k-means.
+
+<!-- livebook:{"break_markdown":true} -->
+
+Before we convert our data to tensor, we will add three more columns which might be informative:
+
+* <pre>rooms_per_family</pre>
+* <pre>bedrooms_per_rooms</pre>
+* <pre>population_per_family</pre>
+
+The names of columns are self-describing. Now, add them to our data frame.
+
+```elixir
+df =
+  Explorer.DataFrame.mutate(df,
+    rooms_per_family: total_rooms / households,
+    bedrooms_per_rooms: total_bedrooms / total_rooms,
+    population_per_family: population / households
+  )
+```
+
+In the next step, we will find the correlation matrix. But to do this, we need to cast our data frame to Nx tensor and split data into train and test sets.
+
+```elixir
+# Module to convert Explorer data frames into tensors.
+defmodule ToTensor do
+  def df_to_matrix(df) do
+    df
+    |> Explorer.DataFrame.names()
+    |> Enum.map(&Explorer.Series.to_tensor(df[&1]))
+    |> Nx.stack(axis: 1)
+  end
+end
+```
+
+```elixir
+# Replace all nils with :nan so we are able to convert to tensor.
+names =
+  df
+  |> Explorer.DataFrame.names()
+  |> List.delete("ocean_proximity")
+
+after_preprocessing =
+  for name <- names, into: %{}, do: {name, Explorer.Series.fill_missing(df[name], :nan)}
+
+preprocessed_data = Explorer.DataFrame.new(after_preprocessing)
+
+# Map ocean_proximity from categorical to float
+# We can consider this categorical as ordinal data since we can order them by the distance
+# to the ocean. The bigger value the further from the ocean.
+mapping = %{
+  "ISLAND" => 0.0,
+  "<1H OCEAN" => 1.0,
+  "NEAR OCEAN" => 2.0,
+  "NEAR BAY" => 3.0,
+  "INLAND" => 4.0
+}
+
+mapped_location =
+  Explorer.Series.transform(df["ocean_proximity"], fn x -> Map.fetch!(mapping, x) end)
+
+df = Explorer.DataFrame.put(preprocessed_data, :ocean_proximity, mapped_location)
+```
+
+```elixir
+# Shuffle data to make splitting more resonable
+{num_rows, _num_cols} = Explorer.DataFrame.shape(df)
+permutation_indices = Nx.iota({num_rows}) |> Nx.shuffle(axis: 0)
+
+y =
+  df[["median_house_value"]]
+  |> ToTensor.df_to_matrix()
+  |> Nx.take(permutation_indices)
+  |> Nx.squeeze()
+
+x =
+  df
+  |> Explorer.DataFrame.discard(["median_house_value"])
+  |> ToTensor.df_to_matrix()
+  |> Nx.take(permutation_indices)
+
+{x, y}
+```
+
+Since we don't have a stratified spilt of data implemented (to learn more see [Stratified Sampling](https://en.wikipedia.org/wiki/Stratified_sampling)), we shuffle data set and take advantage of [Law of large numbers](https://en.wikipedia.org/wiki/Law_of_large_numbers). Now, we will split the data into training and test sets.
```

Comment:
```suggestion
Since we don't have a stratified split of data implemented (to learn more see [Stratified Sampling](https://en.wikipedia.org/wiki/Stratified_sampling)), we shuffle data set and take advantage of [Law of large numbers](https://en.wikipedia.org/wiki/Law_of_large_numbers). Now, we will split the data into training and test sets.
```

Reply:
I would either explain how the law of large numbers comes in play here or just not mention it at all

Reply:
Yeah, I will add more explanation about the law of large numbers. I find it important to explain the sampling process.